File: 1218-lowest-common-ancestor-of-deepest-leaves/lowest-common-ancestor-of-deepest-leaves.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<int,int>mp;
    int maxD;
    void depth(TreeNode* root,int d){
        if(root==NULL){
            return;
        }
        mp[root->val]=d;
        maxD=max(maxD,d);
        depth(root->left,d+1);
        depth(root->right,d+1);
    }
    TreeNode* LCA(TreeNode* root){
        if(root==NULL || mp[root->val]==maxD){
            return root;
        }
        TreeNode* leftAns=LCA(root->left);
        TreeNode* rightAns=LCA(root->right);
        if(leftAns && rightAns) return root;
        // else if(!leftAns && rightAns) return rightAns;
        // else if(leftAns && !rightAns) return leftAns;
        // return nullptr;
        return leftAns!=NULL ? leftAns : rightAns;
    }
    TreeNode* lcaDeepestLeaves(TreeNode* root) {
        depth(root,0);
        return LCA(root);

    }
};

File: 1218-lowest-common-ancestor-of-deepest-leaves/lowest-common-ancestor-of-deepest-leaves.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    class Pair {
        int depth;
        TreeNode node;
        Pair(int _depth,TreeNode _node) {
            this.depth=_depth;
            this.node=_node;
        }
    }
    Pair LCA(TreeNode root){
        if(root==null) {
            return new Pair(0,null);
        }

        Pair leftAns=LCA(root.left);
        Pair rightAns=LCA(root.right);

        if(leftAns.depth==rightAns.depth) return new Pair(leftAns.depth+1,root);
        else if(leftAns.depth > rightAns.depth) return new Pair(leftAns.depth+1,leftAns.node);
        else return new Pair(rightAns.depth+1,rightAns.node);
    }
    public TreeNode lcaDeepestLeaves(TreeNode root) {
        return LCA(root).node;
    }
}